src/System.Private.ServiceModel/tests/Common/Infrastructure/src/BridgeClientCertificateManager.cs

```diff
@@ -0,0 +1,394 @@
+            requestParams.Add(SubjectKeyResourceString, LocalFqdn);
+
+            var response = BridgeClient.MakeResourcePutRequest(MachineCertificateResourceName, requestParams);
+
+            string thumbprint;
+            bool foundLocalCertificate = false;
+
+            lock(s_certificateLock)
+            {
+                if (response.TryGetValue(ThumbprintKeyResourceString, out thumbprint))
+                {
+                    foundLocalCertificate = s_myCertificates.ContainsKey(thumbprint);
+
+                    // The Bridge tells us if the request has been made for a local certificate local to the bridge. 
+                    // If it has, then the Bridge itself has already installed that cert as part of the PUT request
+                    // There's no need for us to do this in the BridgeClient.
+                    string isLocalString;
+                    if (response.TryGetValue(IsLocalKeyResourceString, out isLocalString))
+                    {
+                        bool isLocal = false;
+                        if (bool.TryParse(isLocalString, out isLocal) && isLocal)
+                        {
+                            return;
+                        }
+                    }
+                }
+
+                if (!foundLocalCertificate)
+                {
+                    // GET the cert with thumbprint from the Bridge (returns cert in base64 format)
+                    requestParams = new Dictionary<string, string>();
+                    requestParams.Add(ThumbprintKeyResourceString, thumbprint);
+
+                    string base64Cert = string.Empty;
+                    response = BridgeClient.MakeResourceGetRequest(MachineCertificateResourceName, requestParams);
+
+                    string certificateAsBase64;
+                    if (response.TryGetValue(CertificateKeyResourceString, out certificateAsBase64))
+                    {
+                        certificateToInstall = new X509Certificate2(Convert.FromBase64String(certificateAsBase64));
+                    }
+                    else
+                    {
+                        StringBuilder sb = new StringBuilder();
+                        foreach (var pair in response)
+                        {
+                            sb.AppendFormat("{0}  {1} : {2}", Environment.NewLine, pair.Key, pair.Value);
+                        }
+
+                        throw new Exception(
+                            string.Format("Error retrieving {0} certificate from Bridge. Expected '{1}' key in response. Response contents:{2}{3}",
+                                s_LocalFqdn,
+                                CertificateKeyResourceString,
+                                Environment.NewLine,
+                                sb.ToString()));
+                    }
+                }
+            }
+
+            InstallCertificateToMyStore(certificateToInstall);
+
```

Comment:
These 2 calls are outside the lock.  Any issues with concurrently adding the same cert to the store?